### PR TITLE
rozetka 3 actors added

### PR DESCRIPTION
**Pull request by vladyslav-n**
These are 3 new typescript actors for Hlidac Rozetka Project by Geniusee with edits after previous pull request attempt.

### CODE REVIEW

**Comment by rarous** [comment]
Please do not use custom editorconfig, eslintrc nor prettierrc. Also, please, do not use TypeScript. Actors have to be directly usable without any compilation steps.

**Comment by vladyslav-n** [comment]
Dear Aleš,

I’m a developer at Geniusee and we are partners with Apify.
We usually develop our actors for Apify in typescript with a folder dist included so that the actor doesn’t require any build steps before run. Does it still mean we should change the typescript code to javascript? If it’s crucial for the project requirements, we’ll surely do so.

Sincerely yours,

Vlad
14 апр. 2021 г., 15:53 +0300, Aleš Roubíček ***@***.***>, писал:

**Comment by vladyslav-n** [comment]
Oh, I see Hlidac has dist/ directory in .gitignore, so that I missed that dist directory hasn’t actually been pushed to the repo, sorry for that. I could simply rename it to ‘build/’ so that it will be pushed for sure. In that case will typescript still need to be replaced with javascript? Thanks for your time!

**Comment by vladyslav-n** [comment]
Hey @rarous,
I ported Rozetka actors to javascript, please, check them out.
Have a nice day!

File: actors/rozetka-count/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+# Dockerfile reference, see https://docs.docker.com/engine/reference/builder/
+
+# First, specify the base Docker image. Apify provides the following base images
+# for your convenience:
+#  apify/actor-node-basic (Node.js on Alpine Linux, small and fast image)
+#  apify/actor-node-chrome (Node.js + Chrome on Debian)
+#  apify/actor-node-chrome-xvfb (Node.js + Chrome + Xvfb on Debian)
+# For more information, see https://apify.com/docs/actor#base-images
+# Note that you can use any other image from Docker Hub.
+FROM apify/actor-node-chrome:v0.21.10
```
Comment (metalwarrior665) [inline actors/rozetka-count/Dockerfile]: What is the version 0.21.10 doing here (that is quite old right now)? Let's use the new Dockerfiles actor-node-puppeteer-chrome

File: actors/rozetka-count/src/main.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    utils: { log },
+} = Apify;
+
+Apify.main(async () => {
+    const input = await Apify.getInput();
+    const {
+        maxConcurrency = 100,
+        proxyCountryCode,
+        maxRequestsPerCrawl,
+    } = typeof input === 'object' ? input : {
```
Comment (metalwarrior665) [inline actors/rozetka-count/src/main.js]: typeof null === "object". So I would instead do
const input = (await Apify.getInput()) || defaultInput;

It is also more readable

File: actors/rozetka-count/src/main.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            });
+        },
+    });
+
+    log.info('Starting the crawl.');
+    await crawler.run();
+    log.info('Crawl finished.');
+
+    await Apify.pushData({ OUTPUT: await getOrIncStatsValue() })
+    await Apify.setValue(STATS_KEY, null);
```
Comment (metalwarrior665) [inline actors/rozetka-count/src/main.js]: Why you save a null here?

File: actors/rozetka-count/src/main.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                '#debug': Apify.utils.createRequestDebugInfo(request),
+            });
+        },
+    });
+
+    log.info('Starting the crawl.');
+    await crawler.run();
+    log.info('Crawl finished.');
+
+    await Apify.pushData({ OUTPUT: await getOrIncStatsValue() })
```
Comment (metalwarrior665) [inline actors/rozetka-count/src/main.js]: @zpelechova This is how the dataset should look like?

File: actors/rozetka-count/src/tools.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
+import Apify from 'apify';
+import { NAMED_KV_STORE, STATS_KEY } from './consts.js';
+
+const { utils: { log } } = Apify;
+
+export const purgeStatsStore = async () => {
```
Comment (metalwarrior665) [inline actors/rozetka-count/src/tools.js]: Why we use named stores here? Do we need to store something for in-between runs?

File: actors/rozetka-count/Dockerfile (hunk 1 above)
Comment (metalwarrior665) [reply actors/rozetka-count/Dockerfile]: Actually, since you use cheerio, do we need chrome anywhere? If not, let's use `actor-node`

File: actors/rozetka-daily/src/consts.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
+export const LABELS = {
```
Comment (metalwarrior665) [inline actors/rozetka-daily/src/consts.js]: This second actor looks like most of the code is copy pasted from the first one. That will make it harder for maintenance. I would use a single folder for those and just changed the behaviour via input or env var. cc @rarous

**Review by metalwarrior665 — CHANGES_REQUESTED** [review]
Looks pretty good. I haven't check in-depth everything nor I tested if it works. Please look at the comments. Thanks.

File: actors/rozetka-daily/src/consts.js (hunk 6 above)
Comment (rarous) [reply actors/rozetka-daily/src/consts.js]: Yeah, it look like having separate actors for count functionality is not good idea.

File: actors/rozetka-count/Dockerfile (hunk 1 above)
Comment (vladyslav-n) [reply actors/rozetka-count/Dockerfile]: Yeah, this is weird to me too! I was forced to specify a certain tag for the Docker image, because the automatic code checker here didn't allow me to leave it neither without a tag (e.g. simply **FROM apify/actor-node-chrome**) nor use the "latest" tag. So I specified the only tag I could find on the Docker image web page. What would be better in this case?

File: actors/rozetka-count/src/main.js (hunk 2 above)
Comment (vladyslav-n) [reply actors/rozetka-count/src/main.js]: Yeah, thanks for noticing! Will change that.

File: actors/rozetka-count/Dockerfile (hunk 1 above)
Comment (metalwarrior665) [reply actors/rozetka-count/Dockerfile]: @rarous ?

Comment (rarous) [reply actors/rozetka-count/Dockerfile]: Please, see other actor in this repo for reference

File: actors/rozetka-count/src/main.js (hunk 3 above)
Comment (vladyslav-n) [reply actors/rozetka-count/src/main.js]: I use a named store to preserve the results of the counting in case of migration of the docker container. So that, when the migrated container restarts, it takes the initial count value from the store. That's why I clear this store when the crawling is completed, so that another separate crawling cycle doesn't mix up with the previous actor's run resulting count value. Is there a better solution for this or is there any dangers with this approach?

File: actors/rozetka-daily/src/consts.js (hunk 6 above)
Comment (metalwarrior665) [reply actors/rozetka-daily/src/consts.js]: Well it could be multiple actors but definitely should not be multiple folders. But I will leave that up to you.

File: actors/rozetka-count/src/tools.js (hunk 5 above)
Comment (vladyslav-n) [reply actors/rozetka-count/src/tools.js]: I use a named store to preserve the results of the counting in case of migration of the docker container. So that, when the migrated container restarts, it takes the initial count value from the store. That's why I clear this store when the crawling is completed, so that another separate crawling cycle doesn't mix up with the previous actor's run resulting count value. Is there a better solution for this or is there any dangers with this approach?

File: actors/rozetka-count/Dockerfile (hunk 1 above)
Comment (vladyslav-n) [reply actors/rozetka-count/Dockerfile]: OK, so I'll just change the image to **apify/actor-node:14** as it is in other ones. Thanks!

File: actors/rozetka-daily/src/consts.js (hunk 6 above)
Comment (vladyslav-n) [reply actors/rozetka-daily/src/consts.js]: I guess, I would be the most efficient to do the count of the results and scraping the data in parallel, considering the crawling logic there has to be very similar. I just don't understand clearly how we should handle the output in case of having one actor for both purposes (though there are plenty ways how handle this, for example we could save the results to separate named datasets for count the the daily actors). But the reuse of the code is great, and I'm very glad that I'm allowed to do that here.
So, to sum up, which way should I choose — separate actors with some shared code or a single actor (some more specs about the output should be provided here then)?

File: actors/rozetka-count/src/main.js (hunk 4 above)
Comment (vladyslav-n) [reply actors/rozetka-count/src/main.js]: @zpelechova Since there were no details on the output of the count actor provided in specs, it would be great to know what the correct way of doing this is. I saw an example in another actor as **{ totalCount: value }**. Is this the way it should be done?

File: actors/rozetka-daily/src/consts.js (hunk 6 above)
Comment (rarous) [reply actors/rozetka-daily/src/consts.js]: We have already `common` library for reusable code, but it should be for code that is reusable for all/most of actors. 

This case should be IMHO handled just by `type: "COUNT"` Input parameter. It will be one actor in more modes (we already have this for Black Friday scraping in older actors). It will be scheduled with different input parameters. This mode will just write to Dataset but skip the Keboola upload step.

Sorry, for inconvenience, we are still figuring out the process and shape - count functionality is new requirement for internal benchmarking of scraped data.

File: actors/rozetka-count/src/main.js (hunk 3 above)
Comment (metalwarrior665) [reply actors/rozetka-count/src/main.js]: If it is just for migrated containers, then use the default KV store. Then you don't need to clean it up because it is tied to the run. Named KV is only for saving between multiple runs (migration is still the same run).

File: actors/rozetka-count/src/tools.js (hunk 5 above)
Comment (metalwarrior665) [reply actors/rozetka-count/src/tools.js]: See my other reply :)

File: actors/rozetka-daily/src/consts.js (hunk 6 above)
Comment (metalwarrior665) [reply actors/rozetka-daily/src/consts.js]: I agree with @rarous 's approach, will be the simplest

File: actors/rozetka-count/src/main.js (hunk 3 above)
Comment (vladyslav-n) [reply actors/rozetka-count/src/main.js]: That sounds perfect. Thanks a lot for the advise!

File: actors/rozetka-daily/src/consts.js (hunk 6 above)
Comment (zpelechova) [reply actors/rozetka-daily/src/consts.js]: Hey, I guess it is not properly explained in the docs, but I dont think it makes sense to have an actor four count which does the same as the main actor. The idea behind it is to double check the result, i.e. find and resonably use the numbers which tels how many items there are in each category, like with rozetka here: 
Ноутбуки

**Comment by vladyslav-n** [comment]
@rarous @metalwarrior665 Hi guys!
I suppose, the code is ready for the 2-nd round of the code review)

**Review by metalwarrior665 — APPROVED** [review]
regarding my points, it is done. But @rarous needs to check it works with the system and @zpelechova needs to check the output.

**Comment by janfiedler** [comment]
@vladyslav-n I tried run actor with `"type" = "COUNT"` & `"type" = "DAILY"` and I am getting quite frequently errors:
```
ArgumentError: Expected property string "url" to be a URL, got "/laboratornoe-osnashchenie/c4644808/page=11/" in object "requestLike" at Object.ow [as default] (/Users/janfiedler/Work/TopMonks/GitHub/hlidac-eshopu/node_modules/ow/dist/index.js:19:23)
      at RequestQueue.addRequest (/Users/janfiedler/Work/TopMonks/GitHub/hlidac-eshopu/node_modules/apify/build/storages/request_queue.js:173:21)
      at enqueueLastPage (file:///Users/janfiedler/Work/TopMonks/GitHub/hlidac-eshopu/actors/rozetka-daily/src/routes/helpers/enqueueLastPage.js:17:24)
      at countProductsOrSplitPriceRange (file:///Users/janfiedler/Work/TopMonks/GitHub/hlidac-eshopu/actors/rozetka-daily/src/routes/helpers/countProductsOrSplitPriceRange.js:21:15)
      at handleProductList (file:///Users/janfiedler/Work/TopMonks/GitHub/hlidac-eshopu/actors/rozetka-daily/src/routes/handleProductList.js:43:15)
      at CheerioCrawler.handlePageFunction [as userProvidedHandler] (file:///Users/janfiedler/Work/TopMonks/GitHub/hlidac-eshopu/actors/rozetka-daily/src/main.js:68:27)
      at CheerioCrawler._handleRequestFunction (/Users/janfiedler/Work/TopMonks/GitHub/hlidac-eshopu/node_modules/apify/build/crawlers/cheerio_crawler.js:452:49)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async CheerioCrawler._runTaskFunction (/Users/janfiedler/Work/TopMonks/GitHub/hlidac-eshopu/node_modules/apify/build/crawlers/basic_crawler.js:423:13)
      at async AutoscaledPool._maybeRunTask (/Users/janfiedler/Work/TopMonks/GitHub/hlidac-eshopu/node_modules/apify/build/autoscaling/autoscaled_pool.js:399:13)
```

It is look like, this happen only when part of url is `/page=x/`

**Comment by vladyslav-n** [comment]
@janfiedler Hi!
Seems like there's an issue with relative pathes in urls, will look at it and give some feedback tomorrow.

**Comment by vladyslav-n** [comment]
@janfiedler Hi!
Sorry for the delay. Made some updates to the crawling of the actor and also explicitly added a new allowed content-type to the actor — seems like it failed the actor to work at all in both modes COUNT and DAILY. The crawling changes were made due to some changes in the site categories logic.
